### DISCUSSION TITLE
Adding more authentication backend to tethys

### DISCUSSION THREAD

**Discussion by binhlt0402** [tethysplatform/tethys#1016] [Q&A] [answered]
I've tasked with adding a 3rd party authentication backend and a custom backend for our tethys portal. What is the best way to do it?

**Comment by swainn** [comment]
Hi @binhlt0402 thank you for your question. I'm going to convert this to a discussion and respond there.

**Reply by swainn** [reply, marked as answer]
@binhlt0402 which 3rd party authentication backend do you need to implement? My first recommendation would be to review the backends that are supported out of the box with Tethys here to make sure its not one that we already support: https://docs.tethysplatform.org/en/stable/installation/production/manual/configuration/advanced/social_auth.html

Tethys Platform uses Python Social Auth and the social-auth-app-django extension for Django to implement support for 3rd party auth/single sign on. I would recommend reviewing their extensive list of supported backends: https://python-social-auth.readthedocs.io/en/latest/backends/index.html If you find the backend in that list, you can enable it in Tethys by adding any settings to the `settings` section of your `portal_config.yaml` like so:

```yaml
settings:
  AUTHENTICATION_BACKENDS:
    - social_core.backends.amazon.AmazonOAuth2
  SOCIAL_AUTH_AMAZON_KEY = '...'
  SOCIAL_AUTH_AMAZON_SECRET = '...'
```

If you need to implement your own backend, review the Python Social Auth documentation on doing so: https://python-social-auth.readthedocs.io/en/latest/backends/implementation.html. There are many examples in their repository here: https://github.com/python-social-auth/social-core/tree/master/social_core/backends. Adding your own backend is the same as adding any other backend. Add the class to the `AUTHENTICATION_BACKENDS` list and then add any other settings for your backend to the `portal_config.yaml`

**Comment by binhlt0402** [comment]
Thank you for your answer. It's actually OAuth2 from WSO2. I was thinking of doing it without touching tethys source code. After posting this question I followed the HydroShare backend code and successfully implement our own backend.
